core/src/main/java/org/springframework/security/oauth2/server/authorization/web/ClientCredentialsAuthenticationConverter.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.oauth2.server.authorization.web;
+
+import javax.servlet.http.HttpServletRequest;
+
+import org.springframework.core.convert.converter.Converter;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.oauth2.core.AuthorizationGrantType;
+import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
+import org.springframework.security.oauth2.core.OAuth2Error;
+import org.springframework.security.oauth2.core.OAuth2ErrorCodes;
+import org.springframework.security.oauth2.core.endpoint.OAuth2ParameterNames;
+import org.springframework.security.oauth2.server.authorization.authentication.OAuth2ClientCredentialsAuthenticationToken;
+import org.springframework.util.MultiValueMap;
+import org.springframework.util.StringUtils;
+
+class ClientCredentialsAuthenticationConverter implements Converter<HttpServletRequest, Authentication> {
+
+	@Override
+	public Authentication convert(HttpServletRequest request) {
+		MultiValueMap<String, String> parameters = OAuth2EndpointUtils.getParameters(request);
+
+		// grant_type (REQUIRED)
+		String grantType = parameters.getFirst(OAuth2ParameterNames.GRANT_TYPE);
+		if (!StringUtils.hasText(grantType) ||
+				parameters.get(OAuth2ParameterNames.GRANT_TYPE).size() != 1) {
+			throwError(OAuth2ErrorCodes.INVALID_REQUEST, OAuth2ParameterNames.GRANT_TYPE);
+		}
+		if (!AuthorizationGrantType.CLIENT_CREDENTIALS.getValue().equals(grantType)) {
+			throwError(OAuth2ErrorCodes.UNSUPPORTED_GRANT_TYPE, OAuth2ParameterNames.GRANT_TYPE);
+		}
+
+		// client_id (REQUIRED)
+		String clientId = parameters.getFirst(OAuth2ParameterNames.CLIENT_ID);
+		if (!StringUtils.hasText(clientId)) {
+			throwError(OAuth2ErrorCodes.INVALID_REQUEST, OAuth2ParameterNames.CLIENT_ID);
+		}
+
+		// client_secret (REQUIRED)
+		String clientSecret = parameters.getFirst(OAuth2ParameterNames.CLIENT_SECRET);
+		if (!StringUtils.hasText(clientSecret)) {
+			throwError(OAuth2ErrorCodes.INVALID_REQUEST, OAuth2ParameterNames.CLIENT_SECRET);
+		}
+
+		return new OAuth2ClientCredentialsAuthenticationToken(clientId, clientSecret);
```

Comment:
`scope` is an optional parameter but should also be returned in `OAuth2ClientCredentialsAuthenticationToken`

Reply:
Do you expect section [3.3. Access Token](https://tools.ietf.org/html/rfc6749#section-3.3) to be implemented as a part of this PR? E.g. returning `scope` in the response, and filtering requested scopes based on a list of scopes supported by a client?

Reply:
Yes. The `OAuth2ClientCredentialsAuthenticationProvider` should receive the "request scopes" and the provider should ensure the "request scopes"  are either a subset (or all) of the "assigned scopes". If the client requests a scope that is not in the "assigned scopes" then it should fail. If no scopes are requested then we should auto-approve. However, I'm not sure auto-approving is a good idea but we can change that later if need be.